source/robot_model/src/Model.cpp

```diff
@@ -5,277 +5,381 @@
   }
-  // extract the orientation for the end-effector
-  CartesianState state = cartesian_states.back();
-  delta_r.segment<3>(3 * (cartesian_states.size() - 1)) = state.get_linear_velocity();
-  delta_r.tail(3) = state.get_angular_velocity();
-  jacobian.bottomRows(6) = this->compute_jacobian(joint_state, state.get_name()).data();
-  // compute the Jacobian
-  Eigen::MatrixXd hessian_matrix = jacobian.transpose() * jacobian;
-
-  // set the hessian sparse matrix
-  std::vector<Eigen::Triplet<double>> coefficients;
-  for (unsigned int i = 0; i < nb_joints; ++i) {
-    for (unsigned int j = 0; j < nb_joints; ++j) {
-      coefficients.emplace_back(Eigen::Triplet<double>(i, j, hessian_matrix(i, j)));
+
+  StateRepresentation::JointPositions Model::inverse_geometry(StateRepresentation::JointState actual_joint_state,
+                                                              const StateRepresentation::CartesianState &desired_cartesian_state)
+  {
+    // This method uses the angular velocity evaluated by the 'Infeasibility-free inverse kinematics method' which takes into account the
+    // joint limits. The angular velocity is then used for indicating the direction for an iterative algorithm.
+    // To converge at the desired end-effector position a similar approach to the Newton–Raphson method is used, the only difference is that
+    // here the pseudo-inverse of the jacobian is not used and instead is used the angualr velocity evaluated by the
+    // 'Infeasibility-free inverse kinematics method'.
+
+    const int JOINT_ID = this->robot_model_.njoints - 2;
+
+    // TODO if the actual_joint_state is not provided initialize the position of the robot in a radom conformation
+    // TODO take as input the parameters eps and IT_MAX
+    // TODO return 'success'
+
+    Eigen::VectorXd q = actual_joint_state.get_positions(); //pinocchio::neutral(model);
+    const double eps = 1e-4;
+    const int IT_MAX = 1000;
+
+    std::vector<unsigned int> frame_ids{(unsigned int)JOINT_ID};
+
+    StateRepresentation::CartesianPose actual_cartesian_state;
+    double error;
+
+    bool success = false;
+    Eigen::VectorXd v(this->robot_model_.nv);
+    for (int i = 0;; i++)
+    {
+
+      actual_cartesian_state = this->forward_geometry(actual_joint_state, frame_ids).front();
+      error = desired_cartesian_state.dist(actual_cartesian_state, StateRepresentation::CartesianStateVariable::POSE);
+
+      if (error < eps)
+      {
+        success = true;
+        break;
+      }
+      if (i >= IT_MAX)
+      {
+        success = false;
+        break;
+      }
+      v = this->inverse_kinematic(actual_joint_state, desired_cartesian_state).get_velocities();
+      q = actual_joint_state.get_positions();
+      actual_joint_state.set_velocities(v);
+      StateRepresentation::CartesianTwist x_dot = this->forward_kinematic(actual_joint_state);
+
+      StateRepresentation::CartesianTwist zerovec;
```

Comment:
I'm almost certain that this can be done in one line. @buschbapti ?

Reply:
Yes definitely so here is the code using `JointState` functionalities:

```
JointVelocities dq; = this->inverse_kinematic(current_joint_state, desired_cartesian_twist)
CartesianPose X; = this->forward_geometry(current_joint_state)
CartesianTwist dX; = this->forward_kinematic(dq)
double dt = (desired_cartesian_pose - X).data().norm() / dX.data().norm(); // we could have a norm function on state btw
// you need to convert this dt in a std::chrono but I am not sure about the unit probably milliseconds but the other way is to // assume 1s conversion and divide the state by the double
q += 1s * dq;
q /= dt;
```

Reply:
Right. it is better like you suggest